support data attrs

nwtn opened the pull request:
Fixes #105 by supporting `data-*` for all the TG attributes

This is actually a fairly inelegant commit, and I'm definitely open to suggestions about how to do this more cleanly. I haven't bothered to write new tests yet because I wanted to get a read on what others thought of this approach.

nwtn commented:
test results seem inconsistent and i can't reproduce locally. any help understanding what's happening with this test would be awesome.

nwtn commented:
woo tests passing

nwtn commented:
ping @Shopify/tnt for review

qq99 commented on README.md:
```diff
@@ -57,54 +57,54 @@ You can also tell TurboGraft to refresh the page, but exclude certain elements f
+* `data-refresh-on-error`: (optional) The refresh keys to be refreshed, but using body of XHR which has failed. Only works with error 422. If the XHR returns and error and you do not supply a refresh-on-error, nothing is changed
+* `data-full-refresh-on-error-except`: (optional) Replaces body except for specified refresh keys, using the body of the XHR which has failed.  Only works with error 422
+* `data-remote-once`: (optional) The action will only be performed once. Removes `data-remote-method` and `data-remote-once` from element after consumption
+* `data-full-refresh`: Rather than using the content of the XHR response for partial page replacement, a full page refresh is performed. If `data-refresh-on-success` is defined, the page will be reloaded on these keys. If `data-refresh-on-success` is not defined, a full page refresh is performed. Defaults to true if neither refresh-on-success nor refresh-on-error are provided
```
It's unfortuante that, for whatever reason when I implemented this, I didn't scope them all `tg-*`... :cry:

qq99 commented:
The code looks failproof to me, just `||` conditions

The only thing I'd suggest is that while we're doing it, use `data-tg-*` for everything

nwtn commented:
Namespaced everything. If this approach seems OK I'll do some tests and re-ping shortly.

nwtn commented:
switched back to `getAttribute`, updated tests. @Shopify/tnt

qq99 commented:
:+1: from me

DrewMartin commented:
Nothing in the docs suggests that the old attribute names are still valid so this looks like a breaking change just by the docs. There should probably be something in there about the old names (or make them deprecated?)

DrewMartin commented on lib/assets/javascripts/turbograft/initializers.coffee:
```diff
@@ -4,34 +4,36 @@ hasClass = (node, search) ->
+  throw new Error("Turbograft developer error: You did not provide a URL ('data-tg-#{urlAttribute}' attribute) for data-tg-remote") unless httpUrl
 
-  if target.getAttribute("remote-once")
+  if target.getAttribute("data-tg-remote-once") || target.getAttribute("remote-once")
```
Not a fan of all the `getAttribute('attr') || getAttribute('data-tg-attr')`

How about a `getTGAttribute` method that does
```coffeescript
getTGAttribute(node, attr): ->
  node.getAttribute("data-tg-#{attr}") || node.getAttribute(attr)
```

and then same for remove

DrewMartin replied:
Though because of attributes like `tg-remote` this wouldn't quite work.

Maybe something like
```coffeescript
getTGAttribute(node, attr): ->
  tgAttr = if attr.test[0...3] == 'tg-'
    "data-#{attr}"
  else
    "data-tg-#{attr}"

  node.getAttribute(tgAttr) || node.getAttribute(attr)
```

DrewMartin replied:
Benefits: If we do deprecate the non-data attributes then there's only a couple places that really need to be cleaned up

nwtn replied:
I'm cool with switching to a function like and marking the old attribute names as deprecated in the readme. I'll push that up.

nwtn commented:
bumped the version to 0.2, updated docs to deprecate old attr names, added some methods to clean up all the `||` bits. @DrewMartin @qq99 whaddya think?

DrewMartin commented on lib/assets/javascripts/turbograft/initializers.coffee:
```diff
@@ -1,37 +1,56 @@
+removeTGAttribute = (node, attr) ->
+  tgAttr = tgAttribute(attr)
+  node.removeAttribute(tgAttr)
+  node.removeAttribute(tgAttr)
```
One of these lines should be `node.removeAttribute(attr)`

nwtn replied:
ack good catch!

DrewMartin commented on lib/assets/javascripts/turbograft/initializers.coffee:
```diff
@@ -1,37 +1,56 @@
+  node.removeAttribute(tgAttr)
+  node.removeAttribute(tgAttr)
+
+tgAttributeExists = (node, attr) ->
```
I don't think this method is needed. We can just call `getTGAttribute(node, attr)?`

nwtn replied:
i tried that initially, but it was giving me errors in some of the tests

DrewMartin replied:
And if not, it should probably call `hasAttribute`

qq99 commented on lib/assets/javascripts/turbograft/turbolinks.coffee:
```diff
@@ -215,7 +215,7 @@ class window.Turbolinks
           refreshedNodes.push(newNode)
 
-      else if existingNode.getAttribute("refresh-always") == null
+      else if existingNode.getAttribute("data-tg-refresh-always") == null && existingNode.getAttribute("refresh-always") == null
```
`getTGAttribute`?

DrewMartin replied:
Can we make the attribute methods global and use them here as well?

Bonus to use it in query selectors as well to generate the attribute selector

nwtn commented:
Thanks for the feedback! Will address the remaining bits tomorrow

nwtn commented:
@DrewMartin @qq99 feedback addressed

qq99 commented:
:+1:

nwtn commented:
gave it a test run in core at @qq99’s suggestion and didn't run into any issues!

DrewMartin commented on lib/assets/javascripts/turbograft/remote.coffee:
```diff
@@ -103,7 +103,7 @@ class TurboGraft.Remote
     selector = "input:not([type='reset']):not([type='button']):not([type='submit']):not([type='image']), select, textarea"
     inputs = Array::slice.call(form.querySelectorAll(selector))
-    disabledNodes = Array::slice.call(form.querySelectorAll("[tg-remote-noserialize]"))
+    disabledNodes = Array::slice.call(form.querySelectorAll("[data-tg-remote-noserialize], [tg-remote-noserialize]"))
```
this could use querySelectorAllTGAttribute

DrewMartin commented on lib/assets/javascripts/turbograft/remote.coffee:
```diff
@@ -128,7 +128,7 @@ class TurboGraft.Remote
       return
 
-    unless @initiator.hasAttribute('tg-remote-norefresh')
+    unless @initiator.hasAttribute('data-tg-remote-norefresh') || @initiator.hasAttribute('tg-remote-norefresh')
```
And this could use hasTGAttribute

nwtn commented:
dang thought i got them all. should be fixed now, thanks.

DrewMartin commented:
:ship:
